### ISSUE TITLE
푸시앱의 알림 설정이 캐시되어 다른 회원에게 까지 영향을 줄 수 있을까요??

### ISSUE THREAD

**Issue by wookho** [poesis/xe-supercache#9]
http://xepushapp.com/40526

최근 2번정도 전체 회원의 알림설정 값이 초기화 되는 난감한 상황이 벌어졌습니다.
앱 모듈 개발자분의 답변에 따르면 새로 설치한 앱 처럼 되어 초기설정인 "공지사항" 만 알림을 수신 하는 체크 상태로 되었다는데요..

가설
신규앱 설치자가 앱을 설치합니다.
이때 슈퍼캐시에 의해 앱 의 기본알림 설정 부분에 "공지사항" 만 체크된 상태가 캐시 됩니다.
- 이런 개별 앱들의 설정값들은 regid에 의해서 모든 단말기별로 별도 설정값을 저장하고 있습니다.

이 캐시된 것에 의해 회원들이 로그인 이루어지기 전 캐시된 것에 의해 영향을 받아 새로 설치된 앱으로 인식해서 기존 알림항목 설정이 무시되고 새로운 설치된 앱의 기준인 "공지사항" 만 체크된 값으로 db에 다시 저장되는 현상이 있을 수 있을까요??

최근에 푸시앱을 캐시할지 말지 선택 항목이 생겼던데요. 여기서 푸시앱만 체크를 해제해 두면 캐시의 영향 문제는 아닌 것으로 확인될 수 있는 건가요??(만약 재발한다면...)

사실 앱 사용자들은 자동로그인으로 실 사용시는 거의 로그인 상태를 유지 하기에 전체화면 캐시에 푸시앱을 빼도 상관이 없을 까 하는 개인적인 생각입니다.

자동로그인으로 로그인 되기 바로 직전이나 혹은 앱 설치 후 동기화 하기 전에 비로그인 상태로 푸시앱이 이용될 텐데요. 혹시 이런 시점에서 뭔가 문제를 발생시키는건 아닌지 걱정이 되기도 합니다.

**근데 이 가설은 설득력이 떨어지는게 모든 회원이 짧은 캐시 유지시간  동안 모두 앱을 사용을 시도했다는 것도 말이 안되긴 합니다 ㅜㅡ**

**Comment by wookho** [comment]
혹시 모르니 푸시앱은 전체캐시에서 제외도 하고 푸시앱에서 사용하는 위젯페이지외 빈페이지 두개 전체캐시대상 모듈에서 빼 놓았네요...

**Comment by kijin** [comment]
일부 회원의 설정이 꼬였다면 슈퍼 캐시 탓일지도 모르겠지만, 모든 회원과 기본값이 한꺼번에 초기화되었다면 다른 곳에 원인이 있지 않을까 합니다.

슈퍼 캐시 때문에 오브젝트 캐시 사용량이 급증하다 보니 여러 모듈에서 뜻밖의 버그가 발생하는지도...

**Comment by wookho** [comment]
네. 상황 자체가 난해해서 그냥 가장 가능성이 높은 가설을 세워 놓은 거에요. 거의 대부분 회원이 새글 구독을 하고 있는 상황이라 새글로 단말기로 푸시 알림이 올때 순간적으로 이 전체화면 캐시의 영향을 받았다? 라는 상상까지도 해봅니다. 푸시앱 모듈에서 어떻게 영향을 주는지 제가 지식이 없으니 모두 가설 뿐 입니다. 

일단 영향을 줄만한 것을 차단하는 차원에서 전체화면 캐시에서 푸시앱을 제외하는게 방법이 될 수 있는지 질문 정도의 이슈로 보시면 됩니다. 

캐시를 하지 않는다면 제 가설 자체가 발생하지 않으니 다른 원인을 찾거나 혹은 재발하지 않는 결과도 생기지 않을까 해서요.

**Comment by wookho** [comment]
비로그인시만 캐시된다고 생각한다면 가입직전 앱 설치자에 의해 계속 캐시가 되었을건데요. 결국 앱초기 설치 상태로만 있는 단말기들이 캐시가 계속 되었을거라 생각은 해요. 이 상태들이 특정한 시점에 많은 회원에게 영향르을 주고 일부 회원들은 직접 앱 실행과정에서 영향을 받았을수도.  

전부 소설이니 근거나 그런건 전혀 없어요. 일단 푸시앱 개발자분이 가능성을 제시해 주거나 할때까지 이렇게 지켜보기로 할려구요.

**Comment by kijin** [comment]
단말기별 설정값이 어떻게 전달되는지 알아야 추측이 가능하겠습니다. 우선 푸시앱 사용시 전체화면 캐시가 적용되지 않도록 해두셨으니 이 상태에서도 문제가 생기는지 좀더 지켜보시지요.

**Comment by wookho** [comment]
네. 이상태로 지켜보면서 재발이 없다면 의심했던 부분을 더 살피고 재발한다면 다른 원인으로 판단하고 원인을 찾으면 되겠습니다. 

단말기에 앱을 설치하고 실행(사이트접속)을 하면 단말기가 등록이 됩니다. 이때는 동기화가 이루어지지 않아 회원 정보는 비어 있구요. 초기값으로 공지사항만 푸시받는 세팅으로 등록이 됩니다. db에 단말기정보와 기본알림설정으로 등록이 된 후 회원가입이나 로그인으로 동기화를 하게되면 나머지 추가 항목 댓글 쪽지 멘션 알림이 자동으로 체크되게 되어있습니다. db에 이 정보가 기록되어져 이 정보대로 푸시를 선택적으로 보내주게 됩니다. 

그런데 최근 2번 어떤 이유로 동기화를 마친 회원들의 단말기들이 앱을 방금 설치한 단말기 취급을 받아 알람항목설정이 공자사항만 푸시받는 체크상태로 전환이 된 것입니다.

**Comment by wookho** [comment]
참고가 되실까해서 푸시앱모듈 제작자분의 댓글 중 일부를 남겨 드립니다.

---

일단 정리를 해드리자면 기본 푸시 알림 설정값이 변동되는 로직은 오직 4가지 밖에 없습니다.

1) 앱 푸시 알림 설정은 앱 실행자가 처음 앱을 설치하고 접속하면, 웹뷰에서 xepushapp위젯으로 접속하여 세션처리가 이루어짐과 동시에 앱단에서 firebase regid 정보를 가지고서 서버에 접속합니다. 그러면 푸시앱 모듈에서 post로 넘어온 값을 토대로 기기를 등록하고, 이때 초기값을 넣어주게 되는 거죠. 그리고 각종 접속정보를 앱으로 전달하게 됩니다.

반면 앱 사용자가 두번째로 다시 실행하면, 동일한 방식으로 첫 페이지에서 세션처리한 후 저장된 regid로 post 접속하여 각종 접속정보를 전달받습니다. 여기서는 설정값 변동되는 일은 없습니다.

2) 앱 사용자가 앱 로그인 할 때, 관리자 페이지 설정값대로 기본 푸시 알림설정이 변경됩니다.

3) 그리고 이후 앱 사용자가 앱 설정 화면에서 기본 푸시 알림 설정을 변경할 때, 역시 post값으로 regid값과 변경된 설정값을 넘겨주면 서버는 이를 그대로 처리하여 저장하게 됩니다.

4) 마지막으로 앞서 말씀드린대로 옛날에 설정값 형식을 현재의 형식으로 변동하는 업그레이드를 위해 androidpushapp.txt 파일을 검사하여 설정값을 다음과 같이 변동하게 하였습니다.

true-false-true-true-true-true-true-true

->

newpost_true-newcom_false-mypost_com_true-mycom_com_true-mycom_tail_false-notice_true-message_true-mention_true-recommendation_false

그런데 이 작업은 초기화가 아니라 기존 설정값을 그대로 가져오면서 mycom_com_ 이런식의 접두어만 붙여주는 작업이므로 앞서 말씀드린 것처럼 모든 회원을 초기화하는 오류와는 관련이 없어보입니다.

더욱이 위의 1)또는2)의 작업을 할 때, mysql DB insert과정에서 db오류로 인해서 해당 regid로만 insert되어야 하는데, insert가 아닌 모든 DB의 setting 값을 update해버리는 오류가 아닐까도 생각해봤지만, insert문에서 update가 실행될리가 없기에 이 오류도 가능성이 없어보입니다.

**Comment by kijin** [comment]
페이지 표시 직전에 푸시앱 모듈의 `triggerBeforeDisplay()` 메소드에서 reg_id를 포함하여 여러 가지 데이터를 HTML에 넣어 주고 있습니다.

로그인하지 않은 사용자가 앱을 통해 사이트에 접속할 때 전체화면 캐시된 페이지를 받는다면, 로그인하지 않은 다른 사용자의 reg_id와 각종 설정을 그대로 받게 될 가능성이 있습니다. 캐시된 페이지와 그렇지 않은 페이지를 오간다면 심지어 여러 사용자의 reg_id를 뒤죽박죽으로 받을 수도 있고요.

이 상태에서 설정을 변경하거나, 자동으로 설정을 동기화하는 기능이 있다면 엉뚱한 사용자의 reg_id를 기준으로 DB를 업데이트하게 되겠지요? 만약 빈 값이나 엉뚱한 값으로 넘어온다면 update 쿼리의 reg_id 조건이 오작동하면서 모든 회원에게 일괄적용될지도 모르고요. (XE의 notnull 속성은 그다지 믿을 만한 안전장치가 못 됩니다 ㅡ.ㅡ)

> 푸시앱 모듈에서 post로 넘어온 값을 토대로 기기를 등록하고

특히 이 부분의 의심되는데요, 앱에서 전송하는 정보를 푸시앱 모듈에서 얼마나 철저하게 검증하는지에 따라 아무 문제가 되지 않을 수도 있고 심각한 문제가 될 수도 있습니다.

### 해결방안

만약 위의 원인이 맞다면 슈퍼 캐시와 푸시앱 모듈 사이에 협력이 필요합니다.

로그인하지 않은 상태에서도 반드시 실행되어야 하는 함수가 있다면 전체화면 캐시를 불러온 후 그 함수를 강제로 실행하도록 하거나, 트리거를 제공하여 푸시앱 쪽에서 활용할 수 있도록 해야겠습니다.

예를 들어 슈퍼 캐시 모듈에서 fullpageCacheSet (before), fullpageCacheGet (after) 트리거를 제공한다면 푸시앱에서 전자를 활용하여 캐시 저장시 불필요한 내용을 제거하고, 후자를 활용하여 캐시 로딩 후 다시 넣어주는 방식으로... 성능에 미치는 영향을 최소화하면서도 기능에 지장이 없도록 할 수 있겠습니다.

**Comment by wookho** [comment]
맞습니다. 제가 우려하는 부분입니다. 실제 regid 가 없이 등록된 경우도 한번 보기도 했구요. 굉장히 위험한 상황이 시점과 일치해서 우연하게 다른 단말기의 설정갑이 적용될 수 있을 거라고 추측하고 있어요.

**Comment by wookho** [comment]
개인적인 생각인데요. 앱 사용자는 초기에만 비로그인이고 그 이후 거의 로그인 이용입니다. 자동로그인이죠. 차라리 푸시앱을 블랙리스트로 올리거나 저처럼 푸시앱을 전체캐시 안하는 것으로 유도하는게 좋지 않을까요??

**Comment by kijin** [comment]
네, 문제가 빨리 해결되지 않는다면 정식버전에서는 푸시앱을 전체화면 캐시에서 제외시키는 것을 기본값으로 하겠습니다.

**Comment by wookho** [comment]
슈퍼캐시와는 관련 없는 문제 같습니다. 방금 다시 문제가 발생했네요.

**Comment by kijin** [comment]
푸시앱 사용자를 푸시앱이 아닌 것으로 인식했을 가능성도 있기 때문에 (예: 태블릿을 PC모드로 인식했다든지) 연관성을 100% 배제하기는 힘들겠고요...

일단 전체화면 캐시가 동작중일 때는 푸시앱의 display.before 트리거를 무력화시키도록 하는 코드를 추가해 보았으니 적용후 좀더 지켜봐 주시면 좋겠습니다.

**Comment by wookho** [comment]
저희 사이트에 직접 적용하신거죠 ?

**Comment by kijin** [comment]
아뇨, 깃허브에 커밋만 했습니다.

**Comment by kijin** [comment]
푸시앱은 건드리지 않고 슈퍼 캐시 쪽에서 강제로 트리거를 지워버립니다 ㅋ

**Comment by wookho** [comment]
사실 제가 푸시앱에서 설정과 로그인 등을 할때 머무르게 되는 페이지들을 모두 전체화면캐시에서 제외해 놓았거든요. 이렇게 되면 아마 슈퍼캐시의 영향일 확율은 없을 거 같아요.

**Comment by wookho** [comment]
제가 위에 언급한 푸시앱을 위한 특별한 페이지 3개정도 를 제외한 나머지 일반적인 웹페이지에서 캐싱된 것으로  회원들 단말기가 다른 단말기로 오인될 가능성도 있는건가요? 저는 저 3개 페이지를 제외해 놓았기에 절대 안전할거라 생각하고 있었거든요.

**Comment by kijin** [comment]
어디에 버그가 숨어 있을지 모르니 절대 안전이란 없지요. 슈퍼 캐시를 업데이트한 것과 같은 시기에 갑자기 문제가 생겼다고 하셨으니...

모바일 기기와 푸시앱을 감지하는 로직도 [스카이디자인님 제보](https://www.xetown.com/index.php?mid=lakepark&category=5952&document_srl=421737#comment_425048) 때문에 다시 한 번 점검하고 있습니다. 잠시 후 커밋할 부분까지 적용하시면 슈퍼 캐시 쪽에서 의심되는 부분은 거의 다 제거한 상태일 겁니다.

**Comment by wookho** [comment]
슈퍼캐시 업데이트 2-3일 전에 처음 발생한거에요. 그리고 업데이트 후에 오늘까지 3번이요.

**Comment by wookho** [comment]
XEPUSHK 도 별도로 있는거 아시죠? 킷캣 단말기요.

**Comment by kijin** [comment]
네, XEPUSHK도 user-agent에 XEPUSH가 포함되어 있으므로 푸시앱으로 인식합니다.

슈퍼 캐시 쪽에서 대응할 수 있는 것은 방금 커밋한 부분까지입니다. 이후에도 문제가 생긴다면 트리거도 제공하고 다른 개발자분들이 대응할 수 있도록 멍석을 다 깔아 드렸으니 다른 분들이 해결하셔야...

**Comment by kijin** [comment]
라벨 바꿔 달았습니다 ㅋㅋ

**Comment by wookho** [comment]
신경써주셔서 감사합니다. 방금 커밋까지 적용했구요.
슈퍼캐시의 영향이었으면 좋겠다는 생각이 굴뚝 같지만 제 맘처럼 되지만 않겠죠.
또 재발을 한다면 참 난감한 상황이 되어버리네요. 이게 알림을 못받는 상황으로전환 되는 상황이라 앱을 운영하는 입장에서 앱이 무용지물 되는 상황으로 변하는것이라서요..

지금 발생되면 전체회원 db를 일괄 수정해주고 있네요. 이게 최선이라... 
어떤 원인이던 빨리 원인이 밝혀졌으면 좀 편하게 지낼 듯 합니다. 
많은 부분 살펴봐주셔서 감사합니다.

**Comment by wookho** [comment]
중간 상황 알려드리겠습니다.
해당 이슈의 현상이 발생한건 슈퍼캐시 사이트 적용이후인 11월 13일 입니다.

13일,15일,17일 각 발생했구요.

17일 밤 늦게 슈퍼캐시에서 XE푸시앱을 구별하는 것을 보다 정확하게 할 수 있게 개선해 주신 이후 현재
17일 밤 ~ 22일 오전까지 전혀 발생하지 않고 있습니다.

현재 푸시앱모듈의 경우 각 단말기에 앱버전에 따라 알림설정 로직이 조금씩 다르고하여 어떠한 단말기가 캐시되었을때 설정 어떤 단말기 사용자가 해당시점에서 앱설정을 진입하했을대 해당 문제가 발생하는지 저로서는 파악할 수 있지는 않습니다. 하지만 제 경험적인 판단으로만 보았을때는 분명 웹페이지가 캐시된 상태로 되었을 경우가 고려되지 않은 앱 설정 방식에서 오는 문제라고 잠정 판단하고 있습니다.

푸시앱 개발자분께서는 저희 서버 해킹을 의심하고 계셔서 일부러 푸시앱모듈 삭제 클린설치 등을 하지 않고 모니터링 했습니다. 

사실 이러하나 것이 굉장히 의심된다면 근본적인 해결은 푸시앱에서 이러한 갱신되지 않은 웹페이지에서의 영향을 회피하는 방법을 마련하거나 하는 것일텐데 사용자로서는 좀 아쉬운 면이 많이 있네요. 

일단 중간 상황 알려드리려고 댓글 달았습니다.

**Comment by wookho** [comment]
이문제는 캐시된 화면을 받은 비회원 단말기에서 앱설정이 일어날때 전체회원에게 잘못된 설정이 되는 것으로 판단하는 것이 맞을 것 같습니다. 

현재 슈퍼캐시에서 한 조치 xe푸시앱에는 전체화면캐시 미적용 기능이 완벽해진 이후 17일 늦은밤 ~ 30일 까지 전혀 같은 현상이 재발생 하지 않았습니다.

슈퍼캐시모듈에서 할 조치는 완벽하게 끝난 듯 합니다. 마지막 상황 전달해 드립니다.

**Comment by kijin** [comment]
푸시앱에 전체화면 캐시를 적용하더라도 단말기 설정과 관련된 푸시앱 모듈의 트리거는 작동하지 않도록 하는 패치를 적용했었는데 (d4e165e) 이것까지 테스트해 주시기에는 너무 위험부담이 클까요?

**Comment by wookho** [comment]
네. 회원들 중 일부가 이 알림때문에 클레임을 하는 상태라 더 이상 강제로 알림을 일원화해서 세팅하는 상황이 벌어지면 곤란할 거 같아요. 일단 패치는 다른 사이트에서 사용하는 것 봐서 이상유무를 판단해야 할 상황같아요. 반복적으로 알림을 강제로 띄웠더니 반발이 있어요.
